### PR TITLE
[Feature] KL to init setting in PPO

### PR DESCRIPTION
**Pull request by vmoens**
@tcbegley @apbard

### CODE REVIEW

File: torchrl/objectives/ppo.py [hunk 1]
```diff
@@ -71,6 +73,19 @@ class PPOLoss(LossModule):
             policy and critic will only be trained on the policy loss.
             Defaults to ``False``, ie. gradients are propagated to shared
             parameters for both policy and critic losses.
+        kl_init_coef (scalar, optional): A coefficient for the KL-divergence
+            between the current actor configuration and the initial one.
+            If ``kl_init_coef > 0``, a ``init_actor_params`` attribute
+            will be available and contain the initial configuration of the
+            parameters for the actor. These will be used to return a ``"loss_kl_init"``
+            differentiable entry in the output loss TensorDict that can be used
+            to constain the policy to stay close to its original configuration.
```
Comment (apbard) [inline torchrl/objectives/ppo.py:82]:
```suggestion
            to constrain the policy to stay close to its original configuration.
```

File: torchrl/objectives/ppo.py [hunk 2]
```diff
@@ -71,6 +73,19 @@ class PPOLoss(LossModule):
...
+            Defaults to 0 (ie no loss).
```
Comment (apbard) [inline torchrl/objectives/ppo.py:83]:
```suggestion
            Defaults to 0 (i.e., no loss).
```

File: torchrl/objectives/ppo.py [hunk 3]
```diff
@@ -323,6 +363,18 @@ class ClipPPOLoss(PPOLoss):
...
```
Comment (apbard) [inline torchrl/objectives/ppo.py:373]:
```suggestion
            Defaults to 0 (i.e., no KL component in the loss).
```

File: torchrl/objectives/ppo.py [hunk 4]
```diff
@@ -323,6 +363,18 @@ class ClipPPOLoss(PPOLoss):
             policy and critic will only be trained on the policy loss.
             Defaults to ``False``, ie. gradients are propagated to shared
             parameters for both policy and critic losses.
+        kl_init_coef (scalar, optional): A coefficient for the KL-divergence
+            between the current actor configuration and the initial one.
+            If ``kl_init_coef > 0``, a ``init_actor_params`` attribute
+            will be available and contain the initial configuration of the
+            parameters for the actor. These will be used to return a ``"loss_kl_init"``
+            differentiable entry in the output loss TensorDict that can be used
+            to constain the policy to stay close to its original configuration.
```
Comment (apbard) [inline torchrl/objectives/ppo.py:372]:
```suggestion
            to constrain the policy to stay close to its original configuration.
```

File: torchrl/objectives/ppo.py [hunk 5]
```diff
@@ -71,6 +73,19 @@ class PPOLoss(LossModule):
             policy and critic will only be trained on the policy loss.
             Defaults to ``False``, ie. gradients are propagated to shared
             parameters for both policy and critic losses.
+        kl_init_coef (scalar, optional): A coefficient for the KL-divergence
+            between the current actor configuration and the initial one.
+            If ``kl_init_coef > 0``, a ``init_actor_params`` attribute
+            will be available and contain the initial configuration of the
+            parameters for the actor. These will be used to return a ``"loss_kl_init"``
```
Comment (apbard) [inline torchrl/objectives/ppo.py:80]: > ``init_actor_params`` attribute
            will be available and contain the initial configuration of the
            parameters for the actor
            
not super clear. what do you mean?

File: torchrl/objectives/ppo.py [hunk 6]
```diff
@@ -163,6 +185,14 @@ def __init__(
             warnings.warn(_GAMMA_LMBDA_DEPREC_WARNING, category=DeprecationWarning)
             self.gamma = gamma
 
+    @property
+    def init_actor_params(self):
```
Comment (apbard) [inline torchrl/objectives/ppo.py:189]: maybe `actor_init_params` is better, otherwise seems an intialisation method

File: torchrl/objectives/ppo.py (hunk 5 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:80]: The property won't be available otherwise

File: torchrl/objectives/ppo.py [hunk 7]
```diff
@@ -71,6 +73,19 @@ class PPOLoss(LossModule):
             policy and critic will only be trained on the policy loss.
             Defaults to ``False``, ie. gradients are propagated to shared
             parameters for both policy and critic losses.
+        kl_init_coef (scalar, optional): A coefficient for the KL-divergence
```
Comment (apbard) [inline torchrl/objectives/ppo.py:76]: `kl_init_coef` let think that this is an initial value for kl. what about simply `kl_coef`?

Comment (vmoens) [reply torchrl/objectives/ppo.py:76]: We have other KL coefs we need to distinguish them no?

File: torchrl/objectives/ppo.py (hunk 5 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:80]: Right! I usually don't define docstrings for properties bc I don't expect users to know that smth is a property and not a regular attribute (hence they will never call `help(obj.smth)`)
Can we write it in both places?

**Comment by vmoens** [comment]
I think I addressed all your comments @apbard

File: torchrl/objectives/ppo.py (hunk 7 above)
Comment (apbard) [reply torchrl/objectives/ppo.py]: we need a name that says its the kl vs the initial/reference model. `initial_model_kl_coeff`? its long but I can't think to anything better right now

also what about passing a "config" (aka a dict/dataclass) instead? this enables to implement advanced kl controllers like [here](https://github.com/lvwerra/trl/blob/25fa1bd8803c9998d95a3c38b2bb27a85235ce7b/trl/trainer/ppo_trainer.py#L269)

File: torchrl/objectives/ppo.py (hunk 5 above)
Comment (apbard) [reply torchrl/objectives/ppo.py]: I think here it is not very useful. If you want to keep it I would just write something short as `If kl_init_coef > 0, the init_actor_params will raise`

another option is to make the attribute private.

File: torchrl/objectives/ppo.py (hunk 7 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py]: I changed it to kl_coef, I think it's fine. the other one is named beta and it is used differently. There's no risk of clashing IMO

File: torchrl/objectives/ppo.py [hunk 8]
```diff
@@ -652,7 +722,34 @@ def forward(self, tensordict: TensorDictBase) -> TensorDict:
             loss_critic = self.loss_critic(tensordict)
             td_out.set("loss_critic", loss_critic.mean())
 
+        if self.kl_coef:
+            loss_kl_init = self.kl_coef * self._kl_init(dist, tensordict)
+            td_out.set("loss_kl_init", loss_kl_init.mean())
```
Comment (apbard) [inline torchrl/objectives/ppo.py:727]: I will probably change also this to `kl_loss` or `loss_kl`

**Review by apbard — COMMENTED** [review]
LGTM

File: torchrl/objectives/ppo.py (hunk 8 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:727]: kl_loss no because we need all losses to start by loss by convention
For the KLPENPPOLoss we already have a non-differentiable "kl" entry so it's confusing if we have loss_kl and kl together, one would naturally expect them to be related when they're not.
Maybe `"loss_actor_shift"`?

Comment (apbard) [reply torchrl/objectives/ppo.py:727]: "loss_kl_actor_shift" is nice but my concern is that we have 2 losses that are based on KL. we need to make sure it's clear its the kl computed against:
- ref/initial/starting model
- previous step/iteration model

so would change "kl" entry also

File: torchrl/objectives/ppo.py [hunk 9]
```diff
@@ -454,6 +511,9 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
         if self.critic_coef:
             loss_critic = self.loss_critic(tensordict)
             td_out.set("loss_critic", loss_critic.mean())
+        if self.kl_coef:
+            loss_kl_init = self.kl_coef * self._kl_init(dist, tensordict)
+            td_out.set("loss_kl_init", loss_kl_init.mean())
         td_out.set("ESS", ess.mean() / batch)
         return td_out
```
Comment (apbard) [inline torchrl/objectives/ppo.py:519]: detail what is new/old policy in the following docstring

File: torchrl/objectives/ppo.py (hunk 8 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:727]: kl_to_init
kl_to_prev

Comment (apbard) [reply torchrl/objectives/ppo.py]: kl_wrt_init
kl_wrt_prev
?
otherwise good your version with \_to\_

File: torchrl/objectives/ppo.py (hunk 9 above)
Comment (apbard) [reply torchrl/objectives/ppo.py:519]: KLPENPPOLoss one. (you haven't edit that so couldn't comment on those lines)

File: torchrl/objectives/ppo.py [hunk 10]
```diff
@@ -708,7 +708,7 @@ def forward(self, tensordict: TensorDictBase) -> TensorDict:
         td_out = TensorDict(
             {
                 "loss_objective": -neg_loss.mean(),
-                "kl": kl_val.detach().mean(),
+                "kl_to_prev": kl_val.detach().mean(),
```
Comment (apbard) [inline torchrl/objectives/ppo.py:711]: why is not starting with "loss_*"?

Comment (vmoens) [reply torchrl/objectives/ppo.py:711]: there's a detach just after

Comment (vmoens) [reply torchrl/objectives/ppo.py:711]: true, we could separate that loss too
in this case i did not do it for consistency with clip
such that one can safely do 
```
loss = losses["loss_actor"] + loss["loss_critic"]
```
otherwise the KL module needs one more term.

Comment (vmoens) [reply torchrl/objectives/ppo.py:711]: If we want to change that behaviour it has to be in a separate PR bc it is bc breaking

Comment (vmoens) [reply torchrl/objectives/ppo.py:711]: also for context non-loss values are metadata
we should probably ppt them in a separate nested td

File: torchrl/objectives/ppo.py (hunk 9 above)
Comment (apbard) [reply torchrl/objectives/ppo.py:519]: also is KL(old_policy, new_policy) correct? here they use the opposite:  https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#LL960C28-L960C39

Comment (vmoens) [reply torchrl/objectives/ppo.py:519]: image
From the paper

**Comment by gtnv** [comment]
hey @vmoens, is this still being implemented? if not, would you like to close it?
